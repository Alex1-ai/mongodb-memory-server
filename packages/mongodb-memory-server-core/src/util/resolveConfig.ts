import camelCase from 'camelcase';
import finder from 'find-package-json';
import debug from 'debug';

const log = debug('MongoMS:ResolveConfig');

export enum ResolveConfigVariables {
  DOWNLOAD_DIR = 'DOWNLOAD_DIR',
  PLATFORM = 'PLATFORM',
  ARCH = 'ARCH',
  VERSION = 'VERSION',
  DEBUG = 'DEBUG',
  DOWNLOAD_MIRROR = 'DOWNLOAD_MIRROR',
  DOWNLOAD_URL = 'DOWNLOAD_URL',
  DISABLE_POSTINSTALL = 'DISABLE_POSTINSTALL',
  SYSTEM_BINARY = 'SYSTEM_BINARY',
  MD5_CHECK = 'MD5_CHECK',
  USE_LINUX_LSB_RELEASE = 'USE_LINUX_LSB_RELEASE',
  USE_LINUX_OS_RELEASE = 'USE_LINUX_OS_RELEASE',
  USE_LINUX_ANY_RELEASE = 'USE_LINUX_ANY_RELEASE',
  SKIP_OS_RELEASE = 'SKIP_OS_RELEASE',
  ARCHIVE_NAME = 'ARCHIVE_NAME',
}

export const ENV_CONFIG_PREFIX = 'MONGOMS_';
export const defaultValues = new Map<ResolveConfigVariables, string>([
  // apply app-default values here
  [ResolveConfigVariables.VERSION, '4.0.20'],
]);

/**
 * Set an Default value for an specific key
 * Mostly only used internally (for the "global-x.x" packages)
 * @param key The Key the default value should be assigned to
 * @param value The Value what the default should be
 */
export function setDefaultValue(key: ResolveConfigVariables, value: string): void {
  defaultValues.set(key, value);
}

let packageJsonConfig: {
  [key: string]: string;
} = {};
/**
 * Find the nearest package.json for the provided directory
 * @param directory Set an custom directory to search the config in (default: process.cwd())
 */
export function findPackageJson(directory?: string): void {
  const finderIterator = finder(directory || process.cwd()).next();
  log(`Using package.json at "${finderIterator.filename}"`);
  packageJsonConfig = finderIterator.value?.config?.mongodbMemoryServer ?? {};
}
findPackageJson();

/**
 * Resolve "variableName" with a prefix of "ENV_CONFIG_PREFIX"
 * @param variableName The variable to use
 */
export function resolveConfig(variableName: ResolveConfigVariables): string | undefined {
  return (
    process.env[`${ENV_CONFIG_PREFIX}${variableName}`] ??
    packageJsonConfig[camelCase(variableName)] ??
    defaultValues.get(variableName)
  );
}

export default resolveConfig;

/**
 * Convert "1, on, yes, true" to true (otherwise false)
 * @param env The String / Environment Variable to check
 */
export function envToBool(env: string = ''): boolean {
  return ['1', 'on', 'yes', 'true'].indexOf(env.toLowerCase()) !== -1;
}

// enable debug if "MONGOMS_DEBUG" is true
if (envToBool(resolveConfig(ResolveConfigVariables.DEBUG))) {
  debug.enable('MongoMS:*');
}
