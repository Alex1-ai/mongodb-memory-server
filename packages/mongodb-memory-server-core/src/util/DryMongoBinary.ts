import { promises as fspromises, constants } from 'fs';
import debug from 'debug';
import { envToBool, resolveConfig, ResolveConfigVariables } from './resolveConfig';
import { isNullOrUndefined, pathExists } from './utils';
import * as path from 'path';
import { arch, homedir, platform } from 'os';
import findCacheDir from 'find-cache-dir';
import getOS, { AnyOS, isLinuxOS } from './getos';

const log = debug('MongoMS:DryMongoBinary');

export interface BaseDryMongoBinaryOptions {
  version?: string;
  downloadDir?: string;
  os?: AnyOS;
  arch?: string;
  systemBinary?: string;
}

export interface DryMongoBinaryOptions extends BaseDryMongoBinaryOptions {
  version: NonNullable<BaseDryMongoBinaryOptions['version']>;
}

export interface DryMongoBinaryNameOptions {
  version: NonNullable<DryMongoBinaryOptions['version']>;
  arch: NonNullable<DryMongoBinaryOptions['arch']>;
  os: NonNullable<DryMongoBinaryOptions['os']>;
}

export interface DryMongoBinaryPaths {
  resolveConfig: string;
  legacyHomeCache: string;
  modulesCache: string;
  relative: string;
}

/**
 * Locate an Binary, without downloading / locking
 */
export class DryMongoBinary {
  /**
   * Binaries already found, values are: [Version, Path]
   */
  static binaryCache: Map<string, string> = new Map();
  /**
   * Cache the "getOS" call, so that not much has to be re-executed over and over
   */
  static cachedGetOs: AnyOS;

  /**
   * Try to locate an existing binary
   * @returns The Path to an Binary Found, or undefined
   */
  static async locateBinary(opts: DryMongoBinaryOptions): Promise<string | undefined> {
    log(`locateBinary: Trying to locate Binary for version "${opts.version}"`);
    const useOpts = await this.generateOptions(opts);

    if (!!useOpts.systemBinary) {
      log(`locateBinary: env "SYSTEM_BINARY" was provided with value: "${useOpts.systemBinary}"`);

      const systemReturn = await this.getSystemPath(useOpts.systemBinary);

      if (isNullOrUndefined(systemReturn)) {
        throw new Error(
          `Config option "SYSTEM_BINARY" was provided with value "${useOpts.systemBinary}", but no binary could be found!`
        );
      }

      return systemReturn;
    }

    if (this.binaryCache.has(opts.version)) {
      const binary = this.binaryCache.get(opts.version);
      log(`locateBinary: Requested Version found in cache: "[${opts.version}, ${binary}]"`);

      return binary;
    }

    log('locateBinary: running generateDownloadPath');
    const returnValue = await this.generateDownloadPath(useOpts);

    if (!returnValue[0]) {
      log('locateBinary: could not find an existing binary');

      return undefined;
    }

    log(`locateBinary: found binary at "${returnValue[1]}"`);
    this.binaryCache.set(opts.version, returnValue[1]);

    return returnValue[1];
  }

  /**
   * Generate All required options for the binary name / path generation
   */
  static async generateOptions(
    opts?: DryMongoBinaryOptions
  ): Promise<Required<DryMongoBinaryOptions>> {
    log('generateOptions');
    const defaultVersion = resolveConfig(ResolveConfigVariables.VERSION) ?? '4.0.25';
    const ensuredOpts: DryMongoBinaryOptions = isNullOrUndefined(opts)
      ? { version: defaultVersion }
      : opts;

    if (isNullOrUndefined(this.cachedGetOs)) {
      this.cachedGetOs = await getOS();
    }

    const final: Required<DryMongoBinaryOptions> = {
      version: ensuredOpts.version || defaultVersion,
      downloadDir:
        resolveConfig(ResolveConfigVariables.DOWNLOAD_DIR) || ensuredOpts.downloadDir || '',
      os: ensuredOpts.os ?? this.cachedGetOs,
      arch: ensuredOpts.arch || arch(),
      systemBinary:
        resolveConfig(ResolveConfigVariables.SYSTEM_BINARY) || ensuredOpts.systemBinary || '',
    };

    final.downloadDir = path.dirname((await this.generateDownloadPath(final))[1]);

    return final;
  }

  /**
   * Get the full path with filename
   * @return Absoulte Path with FileName
   */
  static getBinaryName(opts: DryMongoBinaryNameOptions): string {
    log('getBinaryName');
    const addExe = platform() === 'win32' ? '.exe' : '';
    const dist = isLinuxOS(opts.os) ? opts.os.dist : opts.os.os;

    return `mongod-${opts.arch}-${dist}-${opts.version}${addExe}`;
  }

  /**
   * Combine basePath with binaryName
   */
  static combineBinaryName(basePath: string, binaryName: string): string {
    log('combineBinaryName');

    return path.resolve(basePath, binaryName);
  }

  /**
   * Probe if the provided "systemBinary" is an existing path
   * @param systemBinary The Path to probe for an System-Binary
   * @return System Binary path or undefined
   */
  static async getSystemPath(systemBinary: string): Promise<string | undefined> {
    log('getSystempath');
    try {
      await fspromises.access(systemBinary, constants.X_OK); // check if the provided path exists and has the execute bit for current user

      log(`getSystemPath: found system binary path at "${systemBinary}"`);

      return systemBinary; // returns if "access" is successful
    } catch (err) {
      log(`getSystemPath: can't find system binary at "${systemBinary}".\n${err.message}`);
    }

    return undefined;
  }

  /**
   * Generate an "MongoBinaryPaths" object
   *
   * This Function should not hit the FileSystem
   * @return an finished "MongoBinaryPaths" object
   */
  static async generatePaths(
    opts: DryMongoBinaryOptions & DryMongoBinaryNameOptions
  ): Promise<DryMongoBinaryPaths> {
    log('generatePaths');
    const final: DryMongoBinaryPaths = {
      legacyHomeCache: '',
      modulesCache: '',
      relative: '',
      resolveConfig: '',
    };
    const binaryName = this.getBinaryName(opts);
    // Assign "node_modules/.cache" to modulesCache

    // if we're in postinstall script, npm will set the cwd too deep
    // when in postinstall, npm will provide an "INIT_CWD" env variable
    let nodeModulesDLDir = process.env['INIT_CWD'] || process.cwd();
    // as long as "node_modules/mongodb-memory-server*" is included in the path, go the paths up
    while (nodeModulesDLDir.includes(`node_modules${path.sep}mongodb-memory-server`)) {
      nodeModulesDLDir = path.resolve(nodeModulesDLDir, '..', '..');
    }

    const tmpModulesCache = findCacheDir({
      name: 'mongodb-memory-server',
      cwd: nodeModulesDLDir,
    });

    if (!isNullOrUndefined(tmpModulesCache)) {
      final.modulesCache = this.combineBinaryName(path.resolve(tmpModulesCache), binaryName);
    }

    const legacyHomeCache = path.resolve(this.homedir(), '.cache/mongodb-binaries');

    final.legacyHomeCache = this.combineBinaryName(legacyHomeCache, binaryName);

    // Resolve the config value "DOWNLOAD_DIR" if provided, otherwise remove from list
    const resolveConfigValue =
      opts.downloadDir || resolveConfig(ResolveConfigVariables.DOWNLOAD_DIR);

    if (!isNullOrUndefined(resolveConfigValue) && resolveConfigValue.length > 0) {
      log(`generatePaths: resolveConfigValue is not empty`);
      final.resolveConfig = this.combineBinaryName(resolveConfigValue, binaryName);
    }

    // Resolve relative to cwd if no other has been found
    final.relative = this.combineBinaryName(
      path.resolve(process.cwd(), 'mongodb-binaries'),
      binaryName
    );

    return final;
  }

  /**
   * Generate the Path where an Binary will be located
   * @return "boolean" indicating if the binary exists at the provided path, and "string" the path to use for the binary
   */
  static async generateDownloadPath(
    opts: DryMongoBinaryOptions & DryMongoBinaryNameOptions
  ): Promise<[boolean, string]> {
    const preferGlobal = envToBool(resolveConfig(ResolveConfigVariables.PREFER_GLOBAL_PATH));
    log(`generateDownloadPath: Generating Download Path, preferGlobal: "${preferGlobal}"`);
    const paths = await this.generatePaths(opts);

    log('generateDownloadPath: Paths:', paths);

    // Section where paths are probed for an existing binary
    if (await pathExists(paths.resolveConfig)) {
      log(
        `generateDownloadPath: Found binary in resolveConfig (DOWNLOAD_DIR): "${paths.resolveConfig}"`
      );

      return [true, paths.resolveConfig];
    }
    if (await pathExists(paths.legacyHomeCache)) {
      log(`generateDownloadPath: Found binary in legacyHomeCache: "${paths.legacyHomeCache}"`);

      return [true, paths.legacyHomeCache];
    }
    if (await pathExists(paths.modulesCache)) {
      log(`generateDownloadPath: Found binary in modulesCache: "${paths.modulesCache}"`);

      return [true, paths.modulesCache];
    }
    if (await pathExists(paths.relative)) {
      log(`generateDownloadPath: Found binary in relative: "${paths.relative}"`);

      return [true, paths.relative];
    }

    // Section where binary path gets generated when no binary was found
    log(`generateDownloadPath: no existing binary for version "${opts.version}" was found`);

    if (paths.resolveConfig.length > 0) {
      log(`generateDownloadPath: using resolveConfig (DOWNLOAD_DIR) "${paths.resolveConfig}"`);

      return [false, paths.resolveConfig];
    }
    if (preferGlobal && !!paths.legacyHomeCache) {
      log(`generateDownloadPath: using global (preferGlobal) "${paths.legacyHomeCache}"`);

      return [false, paths.legacyHomeCache];
    }
    // this case may not happen, if somehow the cwd gets changed outside of "node_modules" reach
    if (paths.modulesCache.length > 0) {
      log(`generateDownloadPath: using modulesCache "${paths.modulesCache}"`);

      return [false, paths.modulesCache];
    }

    log(`generateDownloadPath: using relative "${paths.relative}"`);

    return [false, paths.relative];
  }

  /**
   * This function is used, because jest just dosnt want "os.homedir" to be mocked
   * if someone can find an way to actually mock this in an test, please change it
   */
  private static homedir(): string {
    return homedir();
  }
}
