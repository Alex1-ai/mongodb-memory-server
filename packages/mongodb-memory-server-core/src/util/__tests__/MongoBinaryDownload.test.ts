import { promises as fspromises } from 'fs';
import md5file from 'md5-file';
import * as mkdirp from 'mkdirp';
import { DryMongoBinary } from '../DryMongoBinary';
import MongoBinaryDownload from '../MongoBinaryDownload';
import MongoBinaryDownloadUrl from '../MongoBinaryDownloadUrl';
import { envName, ResolveConfigVariables } from '../resolveConfig';
import * as utils from '../utils';

jest.mock('md5-file');
jest.mock('mkdirp');

describe('MongoBinaryDownload', () => {
  afterEach(() => {
    delete process.env[envName(ResolveConfigVariables.MD5_CHECK)];
  });

  it('checkMD5 attribute can be set via constructor parameter', () => {
    expect(new MongoBinaryDownload({ checkMD5: true, downloadDir: '/' }).checkMD5).toBe(true);
    expect(new MongoBinaryDownload({ checkMD5: false, downloadDir: '/' }).checkMD5).toBe(false);
  });

  it('if checkMD5 input parameter is missing, then it checks "MONGOMS_MD5_CHECK" environment variable', () => {
    expect(new MongoBinaryDownload({ downloadDir: '/' }).checkMD5).toBe(false);
    process.env[envName(ResolveConfigVariables.MD5_CHECK)] = '1';
    expect(new MongoBinaryDownload({ downloadDir: '/' }).checkMD5).toBe(true);
  });

  it('should use direct download', async () => {
    process.env['yarn_https-proxy'] = '';
    process.env['yarn_proxy'] = '';
    process.env['npm_config_https-proxy'] = '';
    process.env['npm_config_proxy'] = '';
    process.env['https_proxy'] = '';
    process.env['http_proxy'] = '';

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'httpDownload').mockResolvedValue('/tmp/someFile.tgz');
    jest.spyOn(utils, 'pathExists').mockResolvedValue(false);

    await du.download('https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-3.6.3.tgz');
    expect(du.httpDownload).toHaveBeenCalledTimes(1);
    const callArg1 = ((du.httpDownload as jest.Mock).mock.calls[0] as Parameters<
      MongoBinaryDownload['httpDownload']
    >)[1];
    expect(callArg1.agent).toBeUndefined();
  });

  it('should skip download if binary tar exists', async () => {
    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'httpDownload').mockResolvedValue('/tmp/someFile.tgz');
    jest.spyOn(utils, 'pathExists').mockResolvedValue(true);

    await du.download('https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-3.6.3.tgz');

    expect(du.httpDownload).not.toHaveBeenCalled();
  });

  it('should pick up proxy from env vars', async () => {
    process.env['yarn_https-proxy'] = 'http://user:pass@proxy:8080';

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'httpDownload').mockResolvedValue('/tmp/someFile.tgz');
    jest.spyOn(utils, 'pathExists').mockResolvedValue(false);

    await du.download('https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-3.6.3.tgz');
    expect(du.httpDownload).toHaveBeenCalledTimes(1);
    const callArg1 = ((du.httpDownload as jest.Mock).mock.calls[0] as Parameters<
      MongoBinaryDownload['httpDownload']
    >)[1];
    utils.assertion(
      !utils.isNullOrUndefined(callArg1.agent),
      new Error('Expected "callArg1.agent" to be defined')
    );
    // @ts-expect-error because "proxy" if soft-private
    expect(callArg1.agent.proxy.href).toBe('http://user:pass@proxy:8080/');
  });

  it('should not reject unauthorized when npm strict-ssl config is false', async () => {
    // npm sets false config value as empty string in env vars
    process.env['npm_config_strict_ssl'] = '';

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'httpDownload').mockResolvedValue('/tmp/someFile.tgz');
    jest.spyOn(utils, 'pathExists').mockResolvedValue(false);

    await du.download('https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-3.6.3.tgz');
    expect(du.httpDownload).toHaveBeenCalledTimes(1);
    const callArg1 = ((du.httpDownload as jest.Mock).mock.calls[0] as Parameters<
      MongoBinaryDownload['httpDownload']
    >)[1];
    expect(callArg1.rejectUnauthorized).toEqual(false);
  });

  it('should reject unauthorized when npm strict-ssl config is true', async () => {
    // npm sets true config value as string 'true' in env vars
    process.env['npm_config_strict_ssl'] = 'true';

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'httpDownload').mockResolvedValue('/tmp/someFile.tgz');
    jest.spyOn(utils, 'pathExists').mockResolvedValue(false);

    await du.download('https://fastdl.mongodb.org/osx/mongodb-osx-ssl-x86_64-3.6.3.tgz');
    expect(du.httpDownload).toHaveBeenCalledTimes(1);
    const callArg1 = ((du.httpDownload as jest.Mock).mock.calls[0] as Parameters<
      MongoBinaryDownload['httpDownload']
    >)[1];
    expect(callArg1.rejectUnauthorized).toEqual(true);
  });

  it('makeMD5check returns true if md5 of downloaded mongoDBArchive is the same as in the reference result', async () => {
    const someMd5 = 'md5';
    const mongoDBArchivePath = '/some/path';
    const fileWithReferenceMd5 = '/another/path';

    jest.spyOn(fspromises, 'readFile').mockResolvedValueOnce(`${someMd5} fileName`);
    jest.spyOn(md5file, 'sync').mockImplementationOnce(() => someMd5);
    jest.spyOn(fspromises, 'unlink').mockResolvedValue(void 0);

    const du = new MongoBinaryDownload({ downloadDir: '/', checkMD5: true });
    jest.spyOn(du, 'download').mockResolvedValue(fileWithReferenceMd5);
    const urlToMongoDBArchivePath = 'some-url';
    const res = await du.makeMD5check(urlToMongoDBArchivePath, mongoDBArchivePath);

    expect(res).toBe(true);
    expect(du.download).toBeCalledWith(urlToMongoDBArchivePath);
    expect(fspromises.readFile).toBeCalledWith(fileWithReferenceMd5);
    expect(fspromises.unlink).toBeCalledTimes(1);
    expect(md5file.sync).toBeCalledWith(mongoDBArchivePath);
  });

  it('makeMD5check throws an error if md5 of downloaded mongoDBArchive is NOT the same as in the reference result', async () => {
    jest.spyOn(fspromises, 'readFile').mockResolvedValueOnce(`someMD5 fileName`);
    jest.spyOn(md5file, 'sync').mockImplementationOnce(() => 'anotherMD5');

    const du = new MongoBinaryDownload({ downloadDir: '/', checkMD5: true });
    jest.spyOn(du, 'download').mockResolvedValue('');
    await expect(du.makeMD5check('', '')).rejects.toMatchInlineSnapshot(
      `[Error: MongoBinaryDownload: md5 check failed]`
    );
  });

  it('false value of checkMD5 attribute disables makeMD5check validation', async () => {
    jest.spyOn(fspromises, 'readFile').mockResolvedValueOnce(`someMD5 fileName`);
    jest.spyOn(md5file, 'sync').mockImplementationOnce(() => 'anotherMD5');

    const du = new MongoBinaryDownload({ downloadDir: '/', checkMD5: false });
    const result = await du.makeMD5check('', '');
    expect(result).toEqual(undefined);
  });

  it('should return the correct path to binary name (getPath)', async () => {
    const downloadDir = '/path/to/downloadDir';
    jest.spyOn(DryMongoBinary, 'generateOptions').mockResolvedValue({
      arch: 'x64',
      version: '4.0.20',
      downloadDir: downloadDir,
      systemBinary: '',
      os: {
        os: 'linux',
        dist: 'ubuntu',
        release: '14',
      },
    });

    const du = new MongoBinaryDownload({ downloadDir });

    // @ts-expect-error because "getPath" is "protected"
    const path = await du.getPath();
    expect(path).toEqual(`${downloadDir}/mongod-x64-ubuntu-4.0.20`);
  });

  it('should print the download progress (printDownloadProgress)', () => {
    const version = '4.0.20';
    process.stdout.isTTY = false;
    jest.spyOn(console, 'log').mockImplementation(() => void 0);
    jest.spyOn(process.stdout, 'write').mockImplementation(() => true);

    const du = new MongoBinaryDownload({
      downloadDir: '/',
      checkMD5: false,
      version,
      platform: 'linux',
    });
    expect(du.dlProgress.current).toBe(0);
    du.dlProgress.length = 5242880;
    du.dlProgress.totalMb = Math.round((du.dlProgress.length / 1048576) * 10) / 10;

    du.printDownloadProgress({ length: 1048576 });
    expect(console.log).toHaveBeenCalledWith(
      `Downloading MongoDB "${version}": 20% (1mb / ${du.dlProgress.totalMb}mb)\r`
    );
    expect(console.log).toBeCalledTimes(1);
    expect(process.stdout.write).not.toHaveBeenCalled();

    du.printDownloadProgress({ length: 10 });
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(process.stdout.write).not.toHaveBeenCalled();
    expect(du.dlProgress.current).toBe(1048576 + 10);

    process.stdout.isTTY = true;
    du.printDownloadProgress({ length: 0 }, true);
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(process.stdout.write).toHaveBeenCalledWith(
      // its still "20" and "1" because of rounding
      `Downloading MongoDB "${version}": 20% (1mb / ${du.dlProgress.totalMb}mb)\r`
    );
    expect(process.stdout.write).toHaveBeenCalledTimes(1);
  });

  it('should directly return the path if already exists', async () => {
    const binaryPath = '/path/to/binary';
    jest.spyOn(utils, 'pathExists').mockResolvedValue(true);

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest
      .spyOn(du, 'startDownload')
      .mockImplementation(() => fail('Expected this function not to be called'));
    // @ts-expect-error because "getPath" is "protected"
    jest.spyOn(du, 'getPath').mockResolvedValue(binaryPath);

    const returnValue = await du.getMongodPath();
    expect(returnValue).toEqual(binaryPath);
    expect(du.startDownload).not.toHaveBeenCalled();
    expect(
      // @ts-expect-error because "getPath" is "protected"
      du.getPath
    ).toHaveBeenCalledTimes(1);
  });

  it('should return the mongodb archive path (startDownload)', async () => {
    const downloadUrl = 'https://fastdl.mongodb.org/linux/mongod-something-something.tgz';
    const archivePath = '/path/to/archive.tgz';
    jest.spyOn(mkdirp, 'default').mockResolvedValue(void 0);
    jest.spyOn(fspromises, 'access').mockResolvedValue(void 0);
    jest.spyOn(MongoBinaryDownloadUrl.prototype, 'getDownloadUrl').mockResolvedValue(downloadUrl);

    const du = new MongoBinaryDownload({ downloadDir: '/' });
    jest.spyOn(du, 'download').mockResolvedValue(archivePath);
    jest.spyOn(du, 'makeMD5check');

    const returnValue = await du.startDownload();
    expect(returnValue).toEqual(archivePath);
    expect(du.makeMD5check).toHaveBeenCalledWith(`${downloadUrl}.md5`, archivePath);
    expect(du.download).toHaveBeenCalledWith(downloadUrl);
  });

  it('should return the mongodb archive path (startDownload)', async () => {
    const customError = new Error('custom fs error');
    jest.spyOn(mkdirp, 'default').mockResolvedValue(void 0);
    jest.spyOn(fspromises, 'access').mockRejectedValue(customError);
    jest.spyOn(console, 'error').mockImplementation(() => void 0);

    const du = new MongoBinaryDownload({ downloadDir: '/' });

    try {
      await du.startDownload();
      fail('Expected "startDownload" to fail');
    } catch (err) {
      expect(err.message).toEqual(customError.message);
      expect(console.error).toHaveBeenCalledTimes(1);
    }
  });
});
