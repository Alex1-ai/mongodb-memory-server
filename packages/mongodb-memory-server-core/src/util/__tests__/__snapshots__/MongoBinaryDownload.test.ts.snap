// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoBinaryDownload makeMD5check throws an error if md5 of downloaded mongoDBArchive is NOT the same as in the reference result 1`] = `"{\\"binarymd5\\":\\"anotherMD5\\",\\"checkfilemd5\\":\\"someMD5\\"}"`;

exports[`MongoBinaryDownload should passthrough options to MongoBinaryDownloadUrl 1`] = `"/path/to/somewhere/mongod-x64-custom-4.0.0"`;
